File: aa/union_find/1527.py
import atexit
import io
import sys

_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER

@atexit.register
def write():
  sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())


class Union_Find:

    def __init__(this, n, points):
        this.parent = [i for i in range(0, n)]
        this.points = points

    def printa(this):
        print(this.parent)

    def find(this, n):
        root = n
        while root != this.parent[root]:
            root = this.parent[root]

        while n != root:
            next = this.parent[n]
            this.parent[n] = root
            n = next

        return root

    def union(this, a, b):
        root_a = this.find(a)
        root_b = this.find(b)

        if root_a == root_b: return

        if root_a > root_b:
            root_a, root_b = root_b, root_a

        this.parent[root_b] = root_a
        this.points[root_a] += this.points[root_b]



def main():
    while True:
        n, m = list(map(int, input().split()))
        if n == 0 and m == 0:
            return

        points = list(map(int, input().split()))

        my_union = Union_Find(n, points)
        wins = 0
        while m > 0:
            m-=1
            option, a, b = list(map(int, input().split()))

            if option == 1:
                my_union.union(a-1, b-1)
            if option == 2:
                guild_a = my_union.find(a-1)
                guild_b = my_union.find(b-1)
                my_guild = my_union.find(0)

                if my_guild == guild_a and my_union.points[guild_a] > my_union.points[guild_b] :
                    wins+=1
                if my_guild == guild_b and my_union.points[guild_b] > my_union.points[guild_a] :
                    wins+=1

        print(wins)



main()

File: aa/union_find/1527.cpp
#include<bits/stdc++.h>
using namespace std;

int initial_points[100100];

class Union_Find{
  public:
  int * parent;
  int * points;
  int n;

  Union_Find(int n){
    this->n = n;
    this->parent = (int *) malloc( (n+5) * sizeof(int));
    this->points = (int *) malloc( (n+5) * sizeof(int));
    for(int i = 0; i < n ; i++){
      this->parent[i] = i;
      this->points[i] = initial_points[i];
    }
  }

  void printa(){
    for(int i = 0; i < this->n; i++){
      printf("%d ", this->parent[i]);
    }
    printf("\n");

    for(int i = 0; i < this->n; i++){
      printf("%d ", this->points[i]);
    }
    printf("\n\n");
  }

  int find(int a){

    int root = a;
    while(root!=this->parent[root]){
      root = this->parent[root];
    }

    int prox;
    while(a != root){
      prox = this->parent[a];
      this->parent[a] = root;
      a = prox;
    }

    return root;
  }

  int unify(int a, int b){
      int root_a = this->find(a);
      int root_b = this->find(b);

      if(root_a == root_b) return -1;

      if (root_a > root_b)
        swap(root_a, root_b);

      this->parent[root_b]  = this->parent[root_a];
      this->points[root_a] += this->points[root_b];

      return root_a;
  }

};

int main(){
  int option, a, b;

  while(true){
    int n, m;

    scanf("%d %d", &n, &m);
    if(n==0 && m == 0) return 0;


    for(int i = 0; i < n;i++){
      scanf("%d", &initial_points[i]);
    }

    Union_Find my_union = Union_Find(n);

    int wins = 0;
    while(m--){
      scanf("%d %d %d", &option, &a, &b);
      if(option==1){
        my_union.unify(a-1, b-1);
      }
      if(option==2){
        int guild_a = my_union.find(a-1);
        int guild_b = my_union.find(b-1);

        int my_guild = my_union.find(0);

        if(my_guild == guild_a && my_union.points[guild_a] > my_union.points[guild_b])
          wins+=1;

        if(my_guild == guild_b && my_union.points[guild_b] > my_union.points[guild_a])
          wins+=1;
      }
    }

    printf("%d\n", wins);

  }







}
